File: python/WordSearch79.py
# Given an m x n grid of characters board and a string word, return true if word exists in the grid. The word can be
# constructed from letters of sequentially adjacent cells, where adjacent cells are horizontally or vertically
# neighboring. The same letter cell may not be used more than once.
from typing import List


def exist(board: List[List[str]], word: str) -> bool:
    rows, columns, length = len(board), len(board[0]), len(word)
    visited = [[False] * columns for _ in range(rows)]

    def depth_first_search(i, j, index):
        if index == length:
            return True
        if i < 0 or i >= rows or j < 0 or j >= columns or word[index] != board[i][j] or visited[i][j]:
            return False
        visited[i][j] = True
        index += 1
        if depth_first_search(i + 1, j, index) or depth_first_search(i - 1, j, index) or depth_first_search(i, j + 1,
                                                                                                            index) or depth_first_search(
            i, j - 1, index):
            return True
        visited[i][j] = False
        return False

    first = word[0]
    for i in range(rows):
        for j in range(columns):
            if first == board[i][j] and depth_first_search(i, j, 0):
                return True
    return False


if __name__ == "__main__":
    print(exist([["A", "B", "C", "E"], ["S", "F", "C", "S"], ["A", "D", "E", "E"]], "ABCCED"))  # True
    print(exist([["A", "B", "C", "E"], ["S", "F", "C", "S"], ["A", "D", "E", "E"]], "SEE"))  # True
    print(exist([["A", "B", "C", "E"], ["S", "F", "C", "S"], ["A", "D", "E", "E"]], "ABCB"))  # False
    print(exist([["A", "B", "C", "E"], ["S", "F", "E", "S"], ["A", "D", "E", "E"]], "ABCESEEEFS"))  # True

File: java/WordSearch79.java
public class WordSearch79 {
    public static void main(String[] args) {
        char[][] board = {{'A', 'B', 'C', 'E'}, {'S', 'F', 'C', 'S'}, {'A', 'D', 'E', 'E'}};
        String word = "ABCCED";
        System.out.println(exist(board, word));

        word = "SEE";
        System.out.println(exist(board, word));

        word = "ABCB";
        System.out.println(exist(board, word));
    }

    public static boolean exist(char[][] board, String word) {
        int rows = board.length, columns = board[0].length;
        boolean[][] visited = new boolean[rows][columns];
        char first = word.charAt(0);
        for (int i = 0; i < rows; i++)
            for (int j = 0; j < columns; j++)
                if (first == board[i][j] && depthFirstSearch(board, word, visited, rows, columns, i, j, 0)) return true;
        return false;
    }

    private static boolean depthFirstSearch(char[][] board, String word, boolean[][] visited, int rows, int columns, int i, int j, int index) {
        if (index == word.length()) return true;
        if (i < 0 || i >= rows || j < 0 || j >= columns || word.charAt(index) != board[i][j] || visited[i][j])
            return false;
        visited[i][j] = true;
        index++;
        if (depthFirstSearch(board, word, visited, rows, columns, i + 1, j, index) || depthFirstSearch(board, word, visited, rows, columns, i - 1, j, index) || depthFirstSearch(board, word, visited, rows, columns, i, j + 1, index) || depthFirstSearch(board, word, visited, rows, columns, i, j - 1, index))
            return true;
        visited[i][j] = false;
        return false;
    }
}
